src/V3LinkInc.cpp

```diff
@@ -0,0 +1,239 @@
+            m_insStmtp->addHereThisAsNext(newp);
+        } else if (m_insMode == IM_AFTER) {
+            m_insStmtp->addNextHere(newp);
+        } else if (m_insMode == IM_WHILE_PRECOND) {
+            AstWhile* whilep = VN_CAST(m_insStmtp, While);
+            UASSERT_OBJ(whilep, nodep, "Insert should be under WHILE");
+            whilep->addPrecondsp(newp);
+        } else {
+            nodep->v3fatalSrc("Unknown InsertMode");
+        }
+        m_insMode = IM_AFTER;
+        m_insStmtp = newp;
+    }
+
+    // VISITORS
+    virtual void visit(AstWhile* nodep) VL_OVERRIDE {
+        // Special, as statements need to be put in different places
+        // Preconditions insert first just before themselves (the normal
+        // rule for other statement types)
+        m_insStmtp = NULL;  // First thing should be new statement
+        iterateAndNextNull(nodep->precondsp());
+        // Conditions insert first at end of precondsp.
+        m_insMode = IM_WHILE_PRECOND;
+        m_insStmtp = nodep;
+        iterateAndNextNull(nodep->condp());
+        // Body insert just before themselves
+        m_insStmtp = NULL;  // First thing should be new statement
+        iterateAndNextNull(nodep->bodysp());
+        iterateAndNextNull(nodep->incsp());
+        // Done the loop
+        m_insStmtp = NULL;  // Next thing should be new statement
+    }
+    virtual void visit(AstNodeFor* nodep) VL_OVERRIDE {
+        nodep->v3fatalSrc(
+            "For statements should have been converted to while statements in V3Begin.cpp");
+    }
+    virtual void visit(AstNodeStmt* nodep) VL_OVERRIDE {
+        if (!nodep->isStatement()) {
+            iterateChildren(nodep);
+            return;
+        }
+        m_insMode = IM_BEFORE;
+        m_insStmtp = nodep;
+        iterateChildren(nodep);
+        m_insStmtp = NULL;  // Next thing should be new statement
+    }
+    void unsupported_visit(AstNode* nodep) {
+        m_unsupportedHere = true;
+        iterateChildren(nodep);
+        m_unsupportedHere = false;
+    }
+    virtual void visit(AstLogAnd* nodep) VL_OVERRIDE { unsupported_visit(nodep); }
+    void prepost_visit(AstNode* nodep) {
+        // Check if we are underneath a statement
+        if (m_insStmtp == NULL) {
+            prepost_non_stmt_visit(nodep);
+        } else {
+            prepost_stmt_visit(nodep);
+        }
+    }
+    void prepost_non_stmt_visit(AstNode* nodep) {
+        iterateChildren(nodep);
+
+        AstNRelinker linker;
+        AstNode* innerp = nodep->op1p();
+        AstNode* inner_rhsp = innerp->cloneTree(true);
+        AstConst* constp = VN_CAST(nodep->op2p(), Const);
+        AstConst* newconstp = constp->cloneTree(true);
+
+        // Make sure it can be lvalue if it's a VarRef
+        if (VN_IS(innerp, VarRef)) {
+            AstNodeVarRef* vr = VN_CAST(innerp, VarRef);
+            vr->lvalue(true);
+        }
+
+        innerp->unlinkFrBack(&linker);
```

Comment:
```suggestion
        innerp->unlinkFrBack();
```

Only need linker if need to insert something back where this was removed.